### PR TITLE
Fixing iptabes ranges

### PR DESCRIPTION
**Pull request by sbezverk**
With introduction of IPv6 support `OUTBOUND_IP_RANGES_EXCLUDE` and `OUTBOUND_IP_RANGES_INCLUDE` must be treated differently, depending on ipv4 or ipv6 ranges are specified. This PR adds logic to correctly deal with mixed ranges.

Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

### CODE REVIEW

File: tools/deb/istio-iptables.sh [hunk 1]
```diff
@@ -157,6 +181,41 @@ if [ -z "${PROXY_GID}" ]; then
...
+ipv4_ranges_include=()
+if [ "${OUTBOUND_IP_RANGES_INCLUDE}" == "*" ]; then
+   ipv6_ranges_include=("*")
+   ipv4_ranges_include=("*")
+else 
+    for range in "${OUTBOUND_IP_RANGES_INCLUDE[@]}"; do
+        r=${range%$pl}
+        if isValidIP "$r"; then
+            if isIPv4 "$r"; then
+                ipv4_ranges_exclude+=("$range")
```
Comment (tiswanso) [inline tools/deb/istio-iptables.sh]: copy/paste error:  ipv4_ranges_exlude -> ipv4_ranges_include

File: tools/deb/istio-iptables.sh [hunk 2]
```diff
@@ -157,6 +181,41 @@ if [ -z "${PROXY_GID}" ]; then
...
+            else
+                ipv6_ranges_exclude+=("$range")
```
Comment (tiswanso) [inline tools/deb/istio-iptables.sh]: copy/paste error:  ipv6_ranges_exlude -> ipv6_ranges_include

File: tools/deb/istio-iptables.sh [hunk 3]
```diff
@@ -157,6 +181,41 @@ if [ -z "${PROXY_GID}" ]; then
...
 fi
 
+#
+# Since OUTBOUND_IP_RANGES_EXCLUDE could carry ipv4 and ipv6 ranges
+# need to split them in different arrays one for ipv4 and one for ipv6
+# in order to not to fail
+pl='/*'
+ipv6_ranges_exclude=()
+ipv4_ranges_exclude=()
+for range in "${OUTBOUND_IP_RANGES_EXCLUDE[@]}"; do
```
Comment (tiswanso) [inline tools/deb/istio-iptables.sh]: General bash comment:  I think this is the first use in this script of bash arrays.  I think there might be bash version portability problems with those.

File: tools/deb/istio-iptables.sh [hunk 4]
```diff
@@ -454,17 +513,31 @@ if [ -n "${ENABLE_INBOUND_IPV6}" ]; then
...
   
+  # Apply outbound IPv6 exclusions. Must be applied before inclusions.
+  if [ ${#ipv6_ranges_exclude[@]} -gt 0 ]; then
+    for cidr in "${ipv6_ranges_exclude[@]}"; do
+      ip6tables -t nat -A ISTIO_OUTPUT -d "${cidr}" -j RETURN
+    done
+  fi
   # Apply outbound IPv6 inclusions.
-  # TODO Need to figure out differentiation  between IPv4 and IPv6 ranges 
   # for now process only "*"
```
Comment (tiswanso) [inline tools/deb/istio-iptables.sh]: You can remove this comment

Comment (sbezverk) [reply tools/deb/istio-iptables.sh]: done

File: tools/deb/istio-iptables.sh (hunk 2 above)
Comment (sbezverk) [reply tools/deb/istio-iptables.sh]: good catch!! thanks, done

File: tools/deb/istio-iptables.sh (hunk 1 above)
Comment (sbezverk) [reply tools/deb/istio-iptables.sh]: good catch!! thanks, done

File: tools/deb/istio-iptables.sh (hunk 3 above)
Comment (sbezverk) [reply tools/deb/istio-iptables.sh]: This code runs in istio_init container and we do control which image is used. So chances of hitting compatibility issue imho is very low.

Comment (tiswanso) [reply tools/deb/istio-iptables.sh]: Well in k8s env it does, but this script is also used in mesh expansion VMs and we have a mirror version currently in Istio CNI that runs in the host OS.  I'll leave for others to comment, e.g. @costinm 

Would be good to move to go as you mentioned in slack but that's longer term.

**Comment by tiswanso** [comment]
I'm on pto this week so will give my approve vote with those couple copy/paste fixes.  If others are ok with the bash array, I am.

**Comment by incfly** [comment]
This breaks mesh expansion.

The root cause seems to be the isValidIP function implementation.

```bash
function isValidIP() {
   if [ "$1" != "${1#*[0-9].[0-9].[0-9].[0-9]}" ]; then
      true
   elif [ "$1" != "${1#*:[0-9a-fA-F]}" ]; then
      true
   else
      echo "not valid ip $1"
   fi
}
isValidIP "10.3.240.0"
```

output `not valid ip 10.3.240.0`, and the iptable setups stops.

**Comment by sbezverk** [comment]
@incfly please open an issue and assign it to me. I will fix it.. Thanks for reporting it..

**Comment by kyessenov** [comment]
This PR has no tests, so not sure how it got approved. The function to check for IPv4 form is very amenable to unit testing.

**Pull request merged**